src/os/bluestore/Allocator.cc

```diff
@@ -266,50 +275,37 @@ double Allocator::get_fragmentation_score()
   return (ideal - score_sum) / (ideal - terrible);
 }
 
-void Allocator::build_free_state_histogram(
-  size_t alloc_unit, Allocator::FreeStateHistogram& hist)
-{
-  auto num_buckets = hist.size();
-  ceph_assert(num_buckets);
-
-  auto base = free_state_hist_bucket::base;
-  auto base_bits = free_state_hist_bucket::base_bits;
-  auto mux = free_state_hist_bucket::mux;
-  // maximum chunk size we track,
-  // provided by the bucket before the last one
-  size_t max =
-    free_state_hist_bucket::get_max(num_buckets - 2, num_buckets);
-
-  auto iterated_allocation = [&](size_t off, size_t len) {
-    size_t idx;
-    if (len <= base) {
-      idx = 0;
-    } else if (len > max) {
-      idx = num_buckets - 1;
-    } else {
-      size_t most_bit = cbits(uint64_t(len-1)) - 1;
-      idx = 1 + ((most_bit - base_bits) / mux);
-    }
-    ceph_assert(idx < num_buckets);
-    ++hist[idx].total;
+/*************
+* Allocator::FreeStateHistogram
+*************/
+using std::function;
 
-    // now calculate the bucket for the chunk after alignment,
-    // resulting chunks shorter than alloc_unit are discarded
-    auto delta = p2roundup(off, alloc_unit) - off;
-    if (len >= delta + alloc_unit) {
-      len -= delta;
-      if (len <= base) {
-        idx = 0;
-      } else if (len > max) {
-        idx = num_buckets - 1;
-      } else {
-        size_t most_bit = cbits(uint64_t(len-1)) - 1;
-        idx = 1 + ((most_bit - base_bits) / mux);
-      }
-      ++hist[idx].aligned;
-      hist[idx].alloc_units += len / alloc_unit;
-    }
-  };
+void Allocator::FreeStateHistogram::record_extent(uint64_t alloc_unit,
+                                                  uint64_t off,
+                                                  uint64_t len)
+{
+  size_t idx = myTraits._get_p2_size_bucket(len);
+  ceph_assert(idx < buckets.size());
+  ++buckets[idx].total;
 
-  foreach(iterated_allocation);
+  // now calculate the bucket for the chunk after alignment,
+  // resulting chunks shorter than alloc_unit are discarded
+  auto delta = p2roundup(off, alloc_unit) - off;
```

Comment:
It does the same as p2nphase, but as I look at it, it feels like p2nphase is useless, since the p2roundup here is more readable.